// @copilotkit/sdk-js/dist/langgraph/middleware.mjs
import { AIMessage, SystemMessage, createMiddleware } from "langchain";
import * as z from "zod";

//#region src/langgraph/middleware.ts
const createAppContextBeforeAgent = (state, runtime) => {
	const messages = state.messages;
	if (!messages || messages.length === 0) return;
	const appContext = state["copilotkit"]?.context ?? runtime?.context;
	if (!appContext || typeof appContext === "string" && appContext.trim() === "" || typeof appContext === "object" && Object.keys(appContext).length === 0) return;
	const contextMessageContent = `App Context:\n${typeof appContext === "string" ? appContext : JSON.stringify(appContext, null, 2)}`;
	const contextMessagePrefix = "App Context:\n";
	const getContentString = (msg) => {
		if (typeof msg.content === "string") return msg.content;
		if (Array.isArray(msg.content) && msg.content[0]?.text) return msg.content[0].text;
		return null;
	};
	let firstSystemIndex = -1;
	for (let i = 0; i < messages.length; i++) {
		const msg = messages[i];
		const type = msg._getType?.();
		if (type === "system" || type === "developer") {
			if (getContentString(msg)?.startsWith(contextMessagePrefix)) continue;
			firstSystemIndex = i;
			break;
		}
	}
	let existingContextIndex = -1;
	for (let i = 0; i < messages.length; i++) {
		const msg = messages[i];
		const type = msg._getType?.();
		if (type === "system" || type === "developer") {
			if (getContentString(msg)?.startsWith(contextMessagePrefix)) {
				existingContextIndex = i;
				break;
			}
		}
	}
	const contextMessage = new SystemMessage({ content: contextMessageContent });
	let updatedMessages;
	if (existingContextIndex !== -1) {
		updatedMessages = [...messages];
		updatedMessages[existingContextIndex] = contextMessage;
	} else {
		const insertIndex = firstSystemIndex !== -1 ? firstSystemIndex + 1 : 0;
		updatedMessages = [
			...messages.slice(0, insertIndex),
			contextMessage,
			...messages.slice(insertIndex)
		];
	}
	return {
		...state,
		messages: updatedMessages
	};
};
const middlewareInput = {
	name: "CopilotKitMiddleware",
	stateSchema: z.object({ copilotkit: z.object({
		actions: z.array(z.any()),
		context: z.any().optional(),
		interceptedToolCalls: z.array(z.any()).optional(),
		originalAIMessageId: z.string().optional()
	}).optional() }),
	wrapModelCall: async (request, handler) => {
		const frontendTools = request.state["copilotkit"]?.actions ?? [];
		if (frontendTools.length === 0) return handler(request);
		const mergedTools = [...request.tools || [], ...frontendTools];
		return handler({
			...request,
			tools: mergedTools
		});
	},
	beforeAgent: createAppContextBeforeAgent,
	afterAgent: (state) => {
		const interceptedToolCalls = state["copilotkit"]?.interceptedToolCalls;
		const originalMessageId = state["copilotkit"]?.originalAIMessageId;
		if (!interceptedToolCalls?.length || !originalMessageId) return;
		let messageFound = false;
		const updatedMessages = state.messages.map((msg) => {
			if (AIMessage.isInstance(msg) && msg.id === originalMessageId) {
				messageFound = true;
				const existingToolCalls = msg.tool_calls || [];
				return new AIMessage({
					content: msg.content,
					tool_calls: [...existingToolCalls, ...interceptedToolCalls],
					id: msg.id
				});
			}
			return msg;
		});
		if (!messageFound) {
			console.warn(`CopilotKit: Could not find message with id ${originalMessageId} to restore tool calls`);
			return;
		}
		return {
			messages: updatedMessages,
			copilotkit: {
				...state["copilotkit"],
				interceptedToolCalls: void 0,
				originalAIMessageId: void 0
			}
		};
	},
	afterModel: (state) => {
		const frontendTools = state["copilotkit"]?.actions ?? [];
		if (frontendTools.length === 0) return;
		const frontendToolNames = new Set(frontendTools.map((t) => t.function?.name || t.name));
		const lastMessage = state.messages[state.messages.length - 1];
		if (!AIMessage.isInstance(lastMessage) || !lastMessage.tool_calls?.length) return;
		const backendToolCalls = [];
		const frontendToolCalls = [];
		for (const call of lastMessage.tool_calls) if (frontendToolNames.has(call.name)) frontendToolCalls.push(call);
		else backendToolCalls.push(call);
		if (frontendToolCalls.length === 0) return;
		const updatedAIMessage = new AIMessage({
			content: lastMessage.content,
			tool_calls: backendToolCalls,
			id: lastMessage.id
		});
		return {
			messages: [...state.messages.slice(0, -1), updatedAIMessage],
			copilotkit: {
				...state["copilotkit"],
				interceptedToolCalls: frontendToolCalls,
				originalAIMessageId: lastMessage.id
			}
		};
	}
};
const createCopilotKitMiddleware = () => {
	return createMiddleware(middlewareInput);
};
const copilotkitMiddleware = createCopilotKitMiddleware();

//#endregion
export { copilotkitMiddleware };
