// @copilotkit/sdk-js/dist/langchain.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_types = require('./langgraph/types.cjs');
const require_utils = require('./langgraph/utils.cjs');

//#region src/langchain.ts
console.warn("Warning: '@copilotkit/sdk-js/langchain' is deprecated and will be removed in a future release. Please use '@copilotkit/sdk-js/langgraph' instead.");

//#endregion
exports.CopilotKitPropertiesAnnotation = require_types.CopilotKitPropertiesAnnotation;
exports.CopilotKitStateAnnotation = require_types.CopilotKitStateAnnotation;
exports.convertActionToDynamicStructuredTool = require_utils.convertActionToDynamicStructuredTool;
exports.convertActionsToDynamicStructuredTools = require_utils.convertActionsToDynamicStructuredTools;
exports.copilotKitCustomizeConfig = require_utils.copilotkitCustomizeConfig;
exports.copilotKitEmitMessage = require_utils.copilotkitEmitMessage;
exports.copilotKitEmitState = require_utils.copilotkitEmitState;
exports.copilotKitEmitToolCall = require_utils.copilotkitEmitToolCall;
exports.copilotKitExit = require_utils.copilotkitExit;
